File: src/elixir/test/crossimpl_tablebase_test.exs
defmodule CrossImplTablebaseTest do
  @moduledoc """
  Cross-Implementation Tablebase Test Runner

  Runs test cases from spec/tests/tablebase_validation.json to verify
  Elixir tablebase implementation matches the shared spec.

  NOTE: Full tablebase tests are SLOW and skipped by default.
  Run with FULL_TABLEBASE=1 to enable all tests:
    FULL_TABLEBASE=1 mix test test/crossimpl_tablebase_test.exs

  Signed-by: agent #43 claude-sonnet-4 via opencode 20260122T10:58:25
  """

  use ExUnit.Case

  alias Underchex.{Types, Tablebase, Moves}

  # Check for full tablebase mode
  @full_tablebase System.get_env("FULL_TABLEBASE") == "1"

  # ============================================================================
  # Test Suite Loading
  # ============================================================================

  defp load_test_suite do
    spec_path =
      Path.join([
        File.cwd!(),
        "..",
        "..",
        "spec",
        "tests",
        "tablebase_validation.json"
      ])

    spec_path
    |> File.read!()
    |> Jason.decode!()
  end

  defp string_to_color("white"), do: :white
  defp string_to_color("black"), do: :black

  defp string_to_piece_type("king"), do: :king
  defp string_to_piece_type("queen"), do: :queen
  defp string_to_piece_type("knight"), do: :knight
  defp string_to_piece_type("pawn"), do: :pawn
  defp string_to_piece_type("lance"), do: :lance
  defp string_to_piece_type("chariot"), do: :chariot

  defp string_to_lance_variant("A"), do: :a
  defp string_to_lance_variant("B"), do: :b
  defp string_to_lance_variant(_), do: nil

  defp build_board_from_spec(setup) do
    setup["pieces"]
    |> Enum.map(fn placement ->
      piece_type = string_to_piece_type(placement["piece"])
      color = string_to_color(placement["color"])
      variant = placement["variant"] && string_to_lance_variant(placement["variant"])
      piece = Types.new_piece(piece_type, color, variant)
      coord = {placement["q"], placement["r"]}
      {coord, piece}
    end)
    |> Map.new()
  end

  defp wdl_to_string(:win), do: "win"
  defp wdl_to_string(:draw), do: "draw"
  defp wdl_to_string(:loss), do: "loss"
  defp wdl_to_string(:unknown), do: "unknown"

  # ============================================================================
  # Configuration Detection Tests
  # ============================================================================

  describe "Tablebase Configuration Detection (from spec)" do
    test "tb_config_001: KvK configuration detection - two kings only" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {0, -3} => Types.new_piece(:king, :black)
      }

      config = Tablebase.detect_config(board)
      assert config == :kvk
    end

    test "tb_config_002: KQvK configuration detection - white queen" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {2, 0} => Types.new_piece(:queen, :white),
        {0, -3} => Types.new_piece(:king, :black)
      }

      config = Tablebase.detect_config(board)
      assert config == :kqvk
    end

    test "tb_config_003: KQvK configuration detection - black queen" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {2, 0} => Types.new_piece(:queen, :black),
        {0, -3} => Types.new_piece(:king, :black)
      }

      config = Tablebase.detect_config(board)
      # Black having queen still detects as KQvK (color-agnostic)
      assert config == :kqvk
    end

    test "tb_config_004: KLvK configuration detection - lance" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {2, 0} => Types.new_piece(:lance, :white, :a),
        {0, -3} => Types.new_piece(:king, :black)
      }

      config = Tablebase.detect_config(board)
      assert config == :klvk
    end

    test "tb_config_005: KCvK configuration detection - chariot" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {2, 0} => Types.new_piece(:chariot, :white),
        {0, -3} => Types.new_piece(:king, :black)
      }

      config = Tablebase.detect_config(board)
      assert config == :kcvk
    end

    test "tb_config_006: KNvK configuration detection - knight" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {2, 0} => Types.new_piece(:knight, :white),
        {0, -3} => Types.new_piece(:king, :black)
      }

      config = Tablebase.detect_config(board)
      assert config == :knvk
    end

    test "tb_config_007: Complex position not supported by tablebase" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {2, 0} => Types.new_piece(:queen, :white),
        {1, 0} => Types.new_piece(:queen, :white),
        {-1, 0} => Types.new_piece(:queen, :white),
        {0, -3} => Types.new_piece(:king, :black)
      }

      config = Tablebase.detect_config(board)
      # Not a supported configuration
      assert config == nil
    end

    test "all tablebaseConfig tests from spec" do
      suite = load_test_suite()

      config_tests =
        suite["testCases"]
        |> Enum.filter(fn tc -> tc["type"] == "tablebaseConfig" end)

      for tc <- config_tests do
        board = build_board_from_spec(tc["setup"])
        config = Tablebase.detect_config(board)

        if tc["expected"]["supported"] do
          assert config != nil,
                 "#{tc["id"]}: #{tc["description"]} - should be supported"

          if expected_config = tc["expected"]["config"] do
            actual_name = config_name(config)

            assert actual_name == expected_config,
                   "#{tc["id"]}: #{tc["description"]} - expected #{expected_config}, got #{actual_name}"
          end
        else
          # Not supported means nil config
          assert config == nil,
                 "#{tc["id"]}: #{tc["description"]} - should not be supported"
        end
      end
    end
  end

  # ============================================================================
  # WDL Lookup Tests - Fast (KvK only)
  # ============================================================================

  describe "Tablebase WDL Lookup - Fast (KvK)" do
    setup do
      Tablebase.init()
      Tablebase.clear()
      Tablebase.generate(:kvk)
      :ok
    end

    test "tb_wdl_001: KvK is always a draw (white to move)" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {0, -3} => Types.new_piece(:king, :black)
      }

      result = Tablebase.probe(board, :white)

      assert {:ok, entry} = result
      assert entry.wdl == :draw
    end

    test "tb_wdl_002: KvK is draw for black too" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {0, -3} => Types.new_piece(:king, :black)
      }

      result = Tablebase.probe(board, :black)

      assert {:ok, entry} = result
      assert entry.wdl == :draw
    end
  end

  # ============================================================================
  # WDL Lookup Tests - Full (requires FULL_TABLEBASE=1)
  # ============================================================================

  describe "Tablebase WDL Lookup - Full (requires FULL_TABLEBASE=1)" do
    setup do
      if @full_tablebase do
        Tablebase.init()
        Tablebase.clear()
        Tablebase.generate_all()
      end

      :ok
    end

    @tag :slow
    test "tb_wdl_003: KQvK - queen side to move is winning" do
      unless @full_tablebase do
        IO.puts("  Skipping (FULL_TABLEBASE not set)")
        assert true
      else
        board = %{
          {0, 0} => Types.new_piece(:king, :white),
          {2, 0} => Types.new_piece(:queen, :white),
          {0, -4} => Types.new_piece(:king, :black)
        }

        result = Tablebase.probe(board, :white)

        assert {:ok, entry} = result
        assert entry.wdl == :win
      end
    end

    @tag :slow
    test "tb_wdl_004: KQvK - lone king to move is losing" do
      unless @full_tablebase do
        IO.puts("  Skipping (FULL_TABLEBASE not set)")
        assert true
      else
        board = %{
          {0, 0} => Types.new_piece(:king, :white),
          {2, 0} => Types.new_piece(:queen, :white),
          {0, -4} => Types.new_piece(:king, :black)
        }

        result = Tablebase.probe(board, :black)

        assert {:ok, entry} = result
        assert entry.wdl == :loss
      end
    end

    @tag :slow
    test "tb_wdl_006: KNvK - knight alone cannot checkmate (draw)" do
      unless @full_tablebase do
        IO.puts("  Skipping (FULL_TABLEBASE not set)")
        assert true
      else
        board = %{
          {0, 0} => Types.new_piece(:king, :white),
          {2, 0} => Types.new_piece(:knight, :white),
          {0, -3} => Types.new_piece(:king, :black)
        }

        result = Tablebase.probe(board, :white)

        assert {:ok, entry} = result
        assert entry.wdl == :draw
      end
    end

    @tag :slow
    test "tb_wdl_007: KLvK - lance with good position is winning" do
      unless @full_tablebase do
        IO.puts("  Skipping (FULL_TABLEBASE not set)")
        assert true
      else
        board = %{
          {0, 0} => Types.new_piece(:king, :white),
          {0, -2} => Types.new_piece(:lance, :white, :a),
          {0, -4} => Types.new_piece(:king, :black)
        }

        result = Tablebase.probe(board, :white)

        assert {:ok, entry} = result
        assert entry.wdl == :win
      end
    end

    @tag :slow
    test "tb_wdl_008: KCvK - chariot with good position is winning" do
      unless @full_tablebase do
        IO.puts("  Skipping (FULL_TABLEBASE not set)")
        assert true
      else
        board = %{
          {0, 0} => Types.new_piece(:king, :white),
          {2, -2} => Types.new_piece(:chariot, :white),
          {3, -4} => Types.new_piece(:king, :black)
        }

        result = Tablebase.probe(board, :white)

        assert {:ok, entry} = result
        assert entry.wdl == :win
      end
    end

    @tag :slow
    test "all tablebaseWDL tests from spec" do
      unless @full_tablebase do
        IO.puts("  Skipping full WDL tests (FULL_TABLEBASE not set)")
        assert true
      else
        suite = load_test_suite()

        wdl_tests =
          suite["testCases"]
          |> Enum.filter(fn tc -> tc["type"] == "tablebaseWDL" end)

        for tc <- wdl_tests do
          board = build_board_from_spec(tc["setup"])
          turn = string_to_color(tc["setup"]["turn"])

          result = Tablebase.probe(board, turn)

          assert {:ok, entry} = result,
                 "#{tc["id"]}: #{tc["description"]} - should find position in tablebase"

          actual_wdl = wdl_to_string(entry.wdl)
          expected_wdl = tc["expected"]["wdl"]

          assert actual_wdl == expected_wdl,
                 "#{tc["id"]}: #{tc["description"]} - expected WDL=#{expected_wdl}, got WDL=#{actual_wdl}"

          if expected_dtm = tc["expected"]["dtm"] do
            assert entry.dtm == expected_dtm,
                   "#{tc["id"]}: #{tc["description"]} - expected DTM=#{expected_dtm}, got DTM=#{entry.dtm}"
          end
        end
      end
    end
  end

  # ============================================================================
  # Move Suggestion Tests
  # ============================================================================

  describe "Tablebase Move Suggestion" do
    setup do
      Tablebase.init()
      Tablebase.clear()
      Tablebase.generate(:kvk)
      :ok
    end

    test "tb_move_002: KvK - no winning move available (draw)" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {0, -3} => Types.new_piece(:king, :black)
      }

      # For KvK, there's no winning move - it's always a draw
      result = Tablebase.probe(board, :white)

      assert {:ok, entry} = result
      assert entry.wdl == :draw
    end

    @tag :slow
    test "tb_move_001: KQvK - tablebase returns a move that maintains win" do
      unless @full_tablebase do
        IO.puts("  Skipping (FULL_TABLEBASE not set)")
        assert true
      else
        Tablebase.generate(:kqvk)

        board = %{
          {0, 0} => Types.new_piece(:king, :white),
          {2, 0} => Types.new_piece(:queen, :white),
          {0, -4} => Types.new_piece(:king, :black)
        }

        result = Tablebase.probe(board, :white)

        assert {:ok, entry} = result
        assert entry.wdl == :win
        assert entry.best_move != nil

        # Apply the move and verify opponent is losing
        new_board = Moves.apply_move(board, entry.best_move)
        new_result = Tablebase.probe(new_board, :black)

        assert {:ok, new_entry} = new_result
        assert new_entry.wdl == :loss
      end
    end
  end

  # ============================================================================
  # Symmetry Tests
  # ============================================================================

  describe "Tablebase Symmetry (requires FULL_TABLEBASE=1)" do
    setup do
      if @full_tablebase do
        Tablebase.init()
        Tablebase.clear()
        Tablebase.generate(:kqvk)
      end

      :ok
    end

    @tag :slow
    test "tb_symmetric_001: Black queen vs white king - queen side wins" do
      unless @full_tablebase do
        IO.puts("  Skipping (FULL_TABLEBASE not set)")
        assert true
      else
        board = %{
          {0, 4} => Types.new_piece(:king, :white),
          {2, 0} => Types.new_piece(:queen, :black),
          {0, 0} => Types.new_piece(:king, :black)
        }

        result = Tablebase.probe(board, :black)

        assert {:ok, entry} = result
        assert entry.wdl == :win
      end
    end

    @tag :slow
    test "tb_symmetric_002: Black queen vs white king - lone king loses" do
      unless @full_tablebase do
        IO.puts("  Skipping (FULL_TABLEBASE not set)")
        assert true
      else
        board = %{
          {0, 4} => Types.new_piece(:king, :white),
          {2, 0} => Types.new_piece(:queen, :black),
          {0, 0} => Types.new_piece(:king, :black)
        }

        result = Tablebase.probe(board, :white)

        assert {:ok, entry} = result
        assert entry.wdl == :loss
      end
    end
  end

  # ============================================================================
  # Coverage Report
  # ============================================================================

  describe "Tablebase Coverage Report" do
    test "reports tablebase test case coverage" do
      suite = load_test_suite()

      config_tests =
        suite["testCases"]
        |> Enum.filter(fn tc -> tc["type"] == "tablebaseConfig" end)
        |> length()

      wdl_tests =
        suite["testCases"]
        |> Enum.filter(fn tc -> tc["type"] == "tablebaseWDL" end)
        |> length()

      move_tests =
        suite["testCases"]
        |> Enum.filter(fn tc -> tc["type"] == "tablebaseMove" end)
        |> length()

      IO.puts("\n\n=== Tablebase Spec Test Coverage Report (Elixir) ===")
      IO.puts("Configuration detection tests: #{config_tests}")
      IO.puts("WDL lookup tests: #{wdl_tests}")
      IO.puts("Move suggestion tests: #{move_tests}")
      IO.puts("Total tablebase spec tests: #{length(suite["testCases"])}")
      IO.puts("Full tablebase mode: #{if @full_tablebase, do: "ENABLED", else: "disabled"}")
      IO.puts("=====================================================\n")

      assert length(suite["testCases"]) > 0
    end
  end

  # Helper function to convert config atom to string name
  defp config_name(:kvk), do: "KvK"
  defp config_name(:kqvk), do: "KQvK"
  defp config_name(:klvk), do: "KLvK"
  defp config_name(:kcvk), do: "KCvK"
  defp config_name(:knvk), do: "KNvK"
  defp config_name(nil), do: "nil"
end

File: src/typescript/tests/crossimpl-tablebase.test.ts
/**
 * Cross-Implementation Tablebase Test Runner
 * 
 * Runs test cases from spec/tests/tablebase_validation.json to verify
 * TypeScript tablebase implementation matches the shared spec.
 * 
 * NOTE: Full tablebase tests are SLOW and skipped by default.
 * Run with FULL_TABLEBASE=1 to enable all tests.
 * 
 * Signed-by: agent #41 claude-sonnet-4 via amp 20260122T10:32:31
 */

import { describe, test, expect, beforeAll } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';

import {
  detectConfiguration,
  probeTablebase,
  getTablebaseMove,
  initializeTablebases,
  clearTablebases,
  getLoadedTablebases,
  generateTablebase,
  setTablebase,
  TablebaseConfig,
} from '../src/tablebase';

import {
  BoardState,
  Piece,
  Color,
  PieceType,
  LanceVariant,
  coordToString,
} from '../src/types';

import { applyMove, generateAllLegalMoves } from '../src/moves';

// Check if full tablebase tests are enabled
const FULL_TABLEBASE = process.env.FULL_TABLEBASE === '1';

// Test case types from spec
interface PiecePlacement {
  piece: string;
  color: string;
  q: number;
  r: number;
  variant?: string;
}

interface TablebaseConfigTestCase {
  id: string;
  description: string;
  type: 'tablebaseConfig';
  setup: {
    pieces: PiecePlacement[];
  };
  expected: {
    config?: string;
    supported: boolean;
  };
}

interface TablebaseWDLTestCase {
  id: string;
  description: string;
  type: 'tablebaseWDL';
  setup: {
    pieces: PiecePlacement[];
    turn: string;
  };
  expected: {
    wdl: 'win' | 'draw' | 'loss';
    dtm?: number;
  };
}

interface TablebaseMoveTestCase {
  id: string;
  description: string;
  type: 'tablebaseMove';
  setup: {
    pieces: PiecePlacement[];
    turn: string;
  };
  expected: {
    hasMove: boolean;
    preservesWin?: boolean;
  };
}

type TablebaseTestCase = TablebaseConfigTestCase | TablebaseWDLTestCase | TablebaseMoveTestCase;

interface TablebaseTestSuite {
  testCases: TablebaseTestCase[];
}

// Load test cases
function loadTestSuite(): TablebaseTestSuite {
  const specPath = path.join(__dirname, '../../../spec/tests/tablebase_validation.json');
  const content = fs.readFileSync(specPath, 'utf-8');
  return JSON.parse(content) as TablebaseTestSuite;
}

// Convert spec piece to our Piece type
function specToPiece(spec: PiecePlacement): Piece {
  const type = spec.piece as PieceType;
  const color = spec.color as Color;
  const piece: Piece = { type, color };
  
  if (spec.variant) {
    return { ...piece, variant: spec.variant as LanceVariant };
  }
  
  return piece;
}

// Build board from spec setup
function buildBoardFromSpec(setup: { pieces: PiecePlacement[] }): BoardState {
  const board: BoardState = new Map();
  
  for (const placement of setup.pieces) {
    const piece = specToPiece(placement);
    const coord = { q: placement.q, r: placement.r };
    board.set(coordToString(coord), piece);
  }
  
  return board;
}

describe('Cross-Implementation Tablebase Tests (spec/tests/tablebase_validation.json)', () => {
  let testSuite: TablebaseTestSuite;
  
  beforeAll(() => {
    testSuite = loadTestSuite();
    // Initialize tablebases before running tests
    clearTablebases();
    
    // For fast tests, only generate KvK
    // For full tests, generate all tablebases
    if (FULL_TABLEBASE) {
      initializeTablebases();
    } else {
      // Only generate KvK for fast tests
      const kvkConfig: TablebaseConfig = { strongerSide: [], weakerSide: [], name: 'KvK' };
      const kvkTablebase = generateTablebase(kvkConfig);
      setTablebase(kvkTablebase);
    }
  });
  
  describe('Tablebase Configuration Detection', () => {
    test('runs all tablebaseConfig tests from spec', () => {
      const configTests = testSuite.testCases.filter(
        (tc): tc is TablebaseConfigTestCase => tc.type === 'tablebaseConfig'
      );
      
      for (const tc of configTests) {
        const board = buildBoardFromSpec(tc.setup);
        const config = detectConfiguration(board);
        
        if (tc.expected.supported) {
          expect(config, `${tc.id}: ${tc.description} - should be supported`).not.toBeNull();
          if (tc.expected.config) {
            expect(config!.name, `${tc.id}: ${tc.description} - config name`).toBe(tc.expected.config);
          }
        } else {
          // Not supported means either null config or complex position
          // For complex positions, detectConfiguration returns null
          expect(config === null || config.strongerSide.length > 2, 
            `${tc.id}: ${tc.description} - should not be supported`).toBe(true);
        }
      }
    });
    
    test('tb_config_001: KvK detection', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -3 },
        ],
      });
      const config = detectConfiguration(board);
      expect(config).not.toBeNull();
      expect(config!.name).toBe('KvK');
    });
    
    test('tb_config_002: KQvK detection', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'queen', color: 'white', q: 2, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -3 },
        ],
      });
      const config = detectConfiguration(board);
      expect(config).not.toBeNull();
      expect(config!.name).toBe('KQvK');
    });
  });
  
  describe('Tablebase WDL Lookups', () => {
    // Full spec test - only run in FULL_TABLEBASE mode
    test.skipIf(!FULL_TABLEBASE)('runs all tablebaseWDL tests from spec', () => {
      const wdlTests = testSuite.testCases.filter(
        (tc): tc is TablebaseWDLTestCase => tc.type === 'tablebaseWDL'
      );
      
      for (const tc of wdlTests) {
        const board = buildBoardFromSpec(tc.setup);
        const turn = tc.setup.turn as Color;
        
        const result = probeTablebase(board, turn);
        
        expect(result.found, `${tc.id}: ${tc.description} - should find position`).toBe(true);
        expect(result.entry?.wdl, `${tc.id}: ${tc.description} - WDL`).toBe(tc.expected.wdl);
        
        if (tc.expected.dtm !== undefined) {
          expect(result.entry?.dtm, `${tc.id}: ${tc.description} - DTM`).toBe(tc.expected.dtm);
        }
      }
    });
    
    // Fast test - KvK only
    test('tb_wdl_001: KvK is always draw', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -3 },
        ],
      });
      const result = probeTablebase(board, 'white');
      expect(result.found).toBe(true);
      expect(result.entry?.wdl).toBe('draw');
    });
    
    test('tb_wdl_002: KvK is draw for black too', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -3 },
        ],
      });
      const result = probeTablebase(board, 'black');
      expect(result.found).toBe(true);
      expect(result.entry?.wdl).toBe('draw');
    });
    
    // Slow tests - only run in FULL_TABLEBASE mode
    test.skipIf(!FULL_TABLEBASE)('tb_wdl_003: KQvK - queen side is winning', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'queen', color: 'white', q: 2, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -4 },
        ],
      });
      const result = probeTablebase(board, 'white');
      expect(result.found).toBe(true);
      expect(result.entry?.wdl).toBe('win');
    });
    
    test.skipIf(!FULL_TABLEBASE)('tb_wdl_004: KQvK - lone king side is losing', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'queen', color: 'white', q: 2, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -4 },
        ],
      });
      const result = probeTablebase(board, 'black');
      expect(result.found).toBe(true);
      expect(result.entry?.wdl).toBe('loss');
    });
    
    test.skipIf(!FULL_TABLEBASE)('tb_wdl_006: KNvK - knight alone is draw', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'knight', color: 'white', q: 2, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -3 },
        ],
      });
      const result = probeTablebase(board, 'white');
      expect(result.found).toBe(true);
      expect(result.entry?.wdl).toBe('draw');
    });
  });
  
  describe('Tablebase Move Suggestions', () => {
    // Full spec test - only run in FULL_TABLEBASE mode
    test.skipIf(!FULL_TABLEBASE)('runs all tablebaseMove tests from spec', () => {
      const moveTests = testSuite.testCases.filter(
        (tc): tc is TablebaseMoveTestCase => tc.type === 'tablebaseMove'
      );
      
      for (const tc of moveTests) {
        const board = buildBoardFromSpec(tc.setup);
        const turn = tc.setup.turn as Color;
        
        const move = getTablebaseMove(board, turn);
        
        if (tc.expected.hasMove) {
          expect(move, `${tc.id}: ${tc.description} - should have move`).not.toBeNull();
          
          if (tc.expected.preservesWin && move) {
            // Verify the move preserves the win
            const newBoard = applyMove(board, move);
            const newResult = probeTablebase(newBoard, turn === 'white' ? 'black' : 'white');
            expect(newResult.found, `${tc.id}: ${tc.description} - resulting position should be in tablebase`).toBe(true);
            // After our winning move, opponent should be losing
            expect(newResult.entry?.wdl, `${tc.id}: ${tc.description} - opponent should be losing`).toBe('loss');
          }
        } else {
          // For draws, there might not be a "best" move stored
          // This is implementation-dependent, so we just check the position is a draw
          const result = probeTablebase(board, turn);
          expect(result.entry?.wdl, `${tc.id}: ${tc.description} - should be draw if no move`).toBe('draw');
        }
      }
    });
    
    // Fast test - KvK move (no winning move, it's a draw)
    test('tb_move_002: KvK - no winning move available (draw)', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -3 },
        ],
      });
      // For KvK, there's no winning move - it's always a draw
      const result = probeTablebase(board, 'white');
      expect(result.entry?.wdl).toBe('draw');
    });
    
    // Slow test - only run in FULL_TABLEBASE mode
    test.skipIf(!FULL_TABLEBASE)('tb_move_001: KQvK winning position has a winning move', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'queen', color: 'white', q: 2, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: -4 },
        ],
      });
      const move = getTablebaseMove(board, 'white');
      expect(move).not.toBeNull();
      
      // Verify the move preserves the win
      const newBoard = applyMove(board, move!);
      const newResult = probeTablebase(newBoard, 'black');
      expect(newResult.entry?.wdl).toBe('loss');
    });
  });
  
  describe('Tablebase Symmetry Tests', () => {
    // Slow tests - only run in FULL_TABLEBASE mode
    test.skipIf(!FULL_TABLEBASE)('tb_symmetric_001: Black queen vs white king - queen side wins', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 4 },
          { piece: 'queen', color: 'black', q: 2, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: 0 },
        ],
      });
      const result = probeTablebase(board, 'black');
      expect(result.found).toBe(true);
      expect(result.entry?.wdl).toBe('win');
    });
    
    test.skipIf(!FULL_TABLEBASE)('tb_symmetric_002: Black queen vs white king - lone king loses', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 4 },
          { piece: 'queen', color: 'black', q: 2, r: 0 },
          { piece: 'king', color: 'black', q: 0, r: 0 },
        ],
      });
      const result = probeTablebase(board, 'white');
      expect(result.found).toBe(true);
      expect(result.entry?.wdl).toBe('loss');
    });
  });
  
  describe('Spec Test Coverage Report', () => {
    test('reports tablebase test case coverage', () => {
      const configTests = testSuite.testCases.filter(tc => tc.type === 'tablebaseConfig');
      const wdlTests = testSuite.testCases.filter(tc => tc.type === 'tablebaseWDL');
      const moveTests = testSuite.testCases.filter(tc => tc.type === 'tablebaseMove');
      
      console.log('\n=== Tablebase Spec Test Coverage Report ===');
      console.log(`Configuration detection tests: ${configTests.length}`);
      console.log(`WDL lookup tests: ${wdlTests.length}`);
      console.log(`Move suggestion tests: ${moveTests.length}`);
      console.log(`Total tablebase spec tests: ${testSuite.testCases.length}`);
      console.log(`Loaded tablebases: ${getLoadedTablebases().join(', ')}`);
      console.log('============================================\n');
      
      expect(testSuite.testCases.length).toBeGreaterThan(0);
    });
  });
});
